packages/expo-cli/src/commands/upload.ts

```diff
@@ -81,50 +105,38 @@ export default function (program: Command) {
       console.log(`  ${LANGUAGES.join(', ')}`);
       console.log();
     })
-    .asyncActionProjectDir(createUploadAction(IOSUploader, IOS_OPTIONS), { checkConfig: true });
-}
+    // TODO: make this work outside the project directory (if someone passes all necessary options for upload)
+    .asyncActionProjectDir(async (projectDir: string, options: IosPlatformOptions) => {
+      try {
+        // TODO: remove this once we verify `fastlane supply` works on linux / windows
+        checkRuntimePlatform('ios');
 
-function setCommonOptions(command: Command, fileExtension: string) {
-  command
-    .option('--latest', 'uploads the latest build (default)')
-    .option('--id <id>', 'id of the build to upload')
-    .option('--path <path>', `path to the ${fileExtension} file`);
-}
-
-type AnyUploader = any;
-
-function createUploadAction(UploaderClass: AnyUploader, optionKeys: string[]) {
-  return async (
-    projectDir: string,
-    command: AndroidPlatformOptions | IosPlatformOptions
-  ): Promise<void> => {
-    try {
-      ensurePlatformIsSupported();
-      ensureOptionsAreValid(command);
-
-      const options = pick(command, optionKeys);
-      if (UploaderClass.validateOptions) {
-        UploaderClass.validateOptions(options);
+        const args = pick(options, SOURCE_OPTIONS);
+        if (size(args) > 1) {
+          throw new Error(`You have to choose only one of: --path, --id, --latest, --url`);
+        }
+        IOSUploader.validateOptions(options);
+        const uploader = new IOSUploader(projectDir, options);
+        await uploader.upload();
+      } catch (err) {
+        log.error('Failed to upload the standalone app to the app store.');
+        throw err;
       }
-      const uploader = new UploaderClass(projectDir, options);
-      await uploader.upload();
-    } catch (err) {
-      log.error('Failed to upload the standalone app to the app store.');
-      throw err;
-    }
-  };
+    });
 }
 
-function ensurePlatformIsSupported(): void {
+function checkRuntimePlatform(targetPlatform: 'android' | 'ios'): void {
   if (process.platform !== 'darwin') {
-    log.error('Unsupported platform! This feature works on macOS only.');
+    if (targetPlatform === 'android') {
+      log.error('Local Android uploads are only supported on macOS.');
+      log(
+        chalk.bold(
+          'Try --use-submission-service flag to upload your app from Expo servers. This feature is still experimental!'
```

Comment:
```suggestion
          'Try the --use-submission-service flag to upload your app from Expo servers. This feature is behind a flag because it is experimental.'
```